core/src/main/scala/org/apache/spark/ui/scope/RDDOperationGraph.scala

```diff
@@ -181,17 +181,18 @@ private[ui] object RDDOperationGraph extends Logging {
   }
 
   /** Return the dot representation of a subgraph in an RDDOperationGraph. */
-  private def makeDotSubgraph(cluster: RDDOperationCluster, indent: String): String = {
-    val subgraph = new StringBuilder
-    subgraph.append(indent + s"subgraph cluster${cluster.id} {\n")
-    subgraph.append(indent + s"""  label="${cluster.name}";\n""")
+  private def makeDotSubgraph(
+    subgraph: StringBuilder,
+    cluster: RDDOperationCluster,
+    indent: String): StringBuilder = {
+    subgraph.append(indent).append(s"subgraph cluster${cluster.id} {\n")
+    subgraph.append(indent).append( s"""  label="${cluster.name}";\n""")
     cluster.childNodes.foreach { node =>
-      subgraph.append(indent + s"  ${makeDotNode(node)};\n")
+      subgraph.append(indent).append(s"  ${makeDotNode(node)};\n")
     }
     cluster.childClusters.foreach { cscope =>
-      subgraph.append(makeDotSubgraph(cscope, indent + "  "))
+      subgraph.append(makeDotSubgraph(subgraph, cscope, indent + "  "))
```

Comment:
You don't need to append the result of `makeDotSubgraph` right? it's already appending to the `StringBuilder`. I think to avoid this, it could be better to not return the `StringBuilder` at all.

Otherwise yeah looking good. This probably helps to some degree.